### PR TITLE
Refactor MixxxMainWindow.

### PR DESCRIPTION
**Pull request by rryan**
Splits MixxxMainWindow into two pieces: CoreServices and WMainWindow.

This will allow us to iterate on new frontends that do not make use of
QtWidgets / QtGui (such as an Android or iOS port) and generally
enforces a cleaner separation between Mixxx services (provided by
individual manager classes) and the Mixxx UI.

This is also a step towards refactoring the bootup process (which
prompted us to add a loading" launch screen because it's so
slow). With a cleaner concept of "phases" of bootup (construction,
initialize, configure, finalize, destruction) we can start the UI
after construction/initialization and delay the longer tasks
(e.g. connecting audio devices and controllers) or push them onto a
different thread.

Ideally all manager classes and Core will remain buildable with the
QtWidgets module disabled.

### CODE REVIEW

**Comment by rryan** [comment]
This is step 2 (step 1 was WMainMenuBar) on a long-term refactor journey for me. I wrote down some of the goals in this design doc:
http://mixxx.org/wiki/doku.php/mixxx_init_refactor

**Comment by daschuer** [comment]
I have just read the linked document and I like the underlying ideas. The multi-threading approach will help to utilize the n-cores of recent CPUs. 

However, I think it is not a good idea to present the GUI before Mixxx is ready to used. 
This is one of the annoyances (old) Windows compared to Linux, Windows predicts to be ready, but you cannot use it util the HDD has stopped seeking. 

An other general thought is: Multitheading increased the complexity significant. If you aim to reduce the complexity this might be a bad deal. But we will see, when it comes to the implementation.

File: src/main.cpp
Comment (daschuer) [inline src/main.cpp]: // scope for core

File: src/main.cpp
Comment (daschuer) [inline src/main.cpp]: // scope forWMainWindow

File: src/library/autodj/autodjfeature.cpp
Comment (daschuer) [inline src/library/autodj/autodjfeature.cpp]: indent

File: src/widget/wmainwindow.cpp
Comment (daschuer) [inline src/widget/wmainwindow.cpp]: indent

**Comment by daschuer** [comment]
OK, I assume there where nothing changed in the business logic, only moving functions around, right?  I have only added some minor nits. 

The other main issue is the release schedule. We have already published a release date for Mixxx 2.1
for February 2016. The current master version contains a lot of improvements worth a soon release without any further additions. So we should brake out a 2.1 branch and announce the beta phase now. 

( _The alpha seams to be blocked by the broken windows build server though 
@Pegasus-RPG: how is your progress? 
I have played with https://www.appveyor.com/ , will it be an alternative?_  )

These planed refactorings will not have a user impact, but have some degree of regression risk. 
So the should go to a master. The same is true for the GSOC projects.

File: src/library/autodj/autodjfeature.cpp
Comment (rryan) [inline src/library/autodj/autodjfeature.cpp]: Fixed
/me can't wait until we use clang-format so we can stop wasting time on this.

File: src/widget/wmainwindow.cpp
Comment (rryan) [inline src/widget/wmainwindow.cpp]: Fixed

File: src/main.cpp
Comment (rryan) [inline src/main.cpp]: done

File: src/main.cpp
Comment (rryan) [inline src/main.cpp]: Done -- though this seems a bit obvious, no?

**Comment by rryan** [comment]
> OK, I assume there where nothing changed in the business logic, only moving functions around, right? I have only added some minor nits.

Yep, that's right.

**Comment by rryan** [comment]
> The other main issue is the release schedule. We have already published a release date for Mixxx 2.1
> for February 2016. The current master version contains a lot of improvements worth a soon release without any further additions. So we should brake out a 2.1 branch and announce the beta phase now.

I agree this shouldn't go in 2.1.

**Comment by Be-ing** [comment]
@rryan are you interested in picking this up again? Now would be the time to do it. Our startup time has gotten really bad.

**Comment by Be-ing** [comment]
This pull request has not been updated in years so I am closing it to reduce the clutter of open PRs. If you solve the merge conflicts and would like to have it merged in the future, please reopen it. However, considering how many merge conflicts there are, it may be easier to start over from the current master branch.

**Comment by rryan** [comment]
> @rryan are you interested in picking this up again? Now would be the time to do it. Our startup time has gotten really bad.

Yea, I am. I'll re-open when/if I get to it though.

**Comment by rryan** [comment]
Re-based against master, tests are passing on macOS and no shared_ptr leaks occur.

**Comment by Be-ing** [comment]
Thanks for picking up work on this again!

**Comment by Be-ing** [comment]
I think there will be a few small merge conflicts between this and #1705, but it looks like they will be easy to resolve.

**Comment by rryan** [comment]
> I think there will be a few small merge conflicts between this and #1705, but it looks like they will be easy to resolve.

Yea, I tried to make this minimally invasive w.r.t. any changes made to the effects system or library.

File: src/main.cpp
Comment (Be-ing) [inline src/main.cpp]: This introduces a regression. No GUI is shown for about 3 seconds, then when the main window is first shown, the loading progress bar is already over half way.

**Review by Be-ing — CHANGES_REQUESTED** [review]
Thanks for starting to clean this up!

File: src/main.cpp
Comment (Be-ing) [reply src/main.cpp]: Ping. I propose moving the construction of services in the CoreServices class from the constructor to a new method that would get called by WMainWindow. Then have CoreServices emit a signal after each service is started that triggers a progress bar update in WMainWindow.

**Comment by Be-ing** [comment]
I would like to merge this relatively soon before more merge conflicts develop.

**Comment by Be-ing** [comment]
@rryan will you have time to continue this soon? If not I might pick it up.

**Comment by rryan** [comment]
> @rryan will you have time to continue this soon? If not I might pick it up.

Yea, it's on my list after #926.

**Comment by Be-ing** [comment]
Superceeded by #3446

**Pull request closed without merging**
